File: scratch/trace_jsx.js
const fs = require('fs');
const code = fs.readFileSync('components/landing/ConnectPage.tsx', 'utf-8');
const lines = code.split('\n');
let depth = 0;
let stack = [];

for(let i = 520; i < lines.length; i++) {
  const line = lines[i];
  
  // A crude regex to find opening and closing tags on the line
  const tagRegex = /<\/?([a-zA-Z0-9\.]+)[^>]*>/g;
  let match;
  while ((match = tagRegex.exec(line)) !== null) {
    const fullTag = match[0];
    const tagName = match[1];
    
    // Ignore self-closing tags like <img />, <Loader2 />, <svg ... /> (we assume svg doesn't have nested divs for this)
    if (fullTag.endsWith('/>')) continue;
    // Ignore br, hr, input etc if any
    
    if (fullTag.startsWith('</')) {
      depth--;
      const popped = stack.pop();
      console.log(String(i+1).padStart(4), depth, " ".repeat(Math.max(0, depth)) + "CLOSE: " + fullTag + " (matched: " + popped + ")");
    } else {
      stack.push(tagName);
      console.log(String(i+1).padStart(4), depth, " ".repeat(Math.max(0, depth)) + "OPEN:  " + fullTag);
      depth++;
    }
  }
}

File: scratch/trace_jsx.py
import re
import sys

def analyze_jsx():
    with open('components/landing/ConnectPage.tsx', 'r', encoding='utf-8') as f:
        lines = f.readlines()
        
    depth = 0
    stack = []
    
    # Simple regex to match tags. Ignores attributes for counting.
    # It catches <div, <motion.div, </div, </motion.div
    tag_pattern = re.compile(r'<(/?(?:div|motion\.div))[^>]*>')
    
    for i in range(523, len(lines)):
        line = lines[i]
        
        # Strip comments to avoid matching commented out tags
        line_no_comments = re.sub(r'\{/\*.*?\*/\}', '', line)
        line_no_comments = re.sub(r'//.*', '', line_no_comments)
        
        matches = tag_pattern.finditer(line_no_comments)
        for match in matches:
            full_tag = match.group(0)
            tag_name = match.group(1)
            
            if full_tag.endswith('/>'):
                continue
                
            if tag_name.startswith('/'):
                if len(stack) == 0:
                    print(f"Error: Found closing tag {tag_name} but stack is empty at line {i+1}")
                    return
                popped = stack.pop()
                depth -= 1
                print(f"{i+1:4d} | Depth: {depth:2d} | CLOSE: {tag_name} (Matches {popped})")
                
                if depth == 0:
                    print(f"ROOT CLOSED AT LINE {i+1}")
            else:
                stack.append(tag_name)
                depth += 1
                print(f"{i+1:4d} | Depth: {depth:2d} | OPEN: {tag_name}")

if __name__ == '__main__':
    analyze_jsx()
